### PR TITLE
Makes Astro.resolve return root-relative paths

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Instead of returning a pathname that is the full filesystem URL this will not return `/src/images/penguin.png`, projectRoot relative.
- Depends on https://github.com/withastro/astro-compiler/pull/179
- Closes https://github.com/withastro/astro/issues/2002
- Closes https://github.com/withastro/astro/issues/1968
- Closes https://github.com/withastro/astro/issues/2068

## Testing

No tests needed to be updated, but some code needed to be updated to make them all pass.

## Docs

Bug fix.

### CODE REVIEW

File: packages/astro/src/runtime/server/index.ts
Comment (natemoo-re) [inline packages/astro/src/runtime/server/index.ts:296]: Do we want to unconditionally `substr` this? What happens if the resolved path is outside the project root?

Comment (natemoo-re) [reply packages/astro/src/runtime/server/index.ts:296]: I'd feel more comfortable with a `replace` or `startsWith` check

**Review by natemoo-re — COMMENTED** [review]
LGTM with one small comment about `substr` vs `replace`

File: packages/astro/src/runtime/server/index.ts
Comment (matthewp) [reply packages/astro/src/runtime/server/index.ts:296]: Yeah, the only time it should be outside of the project root is if someone docs `Astro.resolve('/images/penguin.png')` which would be unexpected, but seems reasonable to support so will change.

Comment (matthewp) [reply packages/astro/src/runtime/server/index.ts]: done!

**Comment by drwpow** [comment]
After merging #2052, that may affect this because of a minor change to `viteifyURL`. Could you rebase off latest `main` and try tests? If anything is failing I can help resolve.

File: packages/astro/src/vite-plugin-build-html/index.ts
Comment (drwpow) [inline packages/astro/src/vite-plugin-build-html/index.ts:146]:
```suggestion
                astroAssetMap.set(src, fs.readFile(resolved));
```
with the change to viteifyURL this may be better

File: packages/astro/src/vite-plugin-build-html/index.ts
Comment (drwpow) [inline packages/astro/src/vite-plugin-build-html/index.ts:145]: This may need a different function than `viteifyURL`. Because this is Rollup, this may need `slash(fileURLToPath(resolved))` which _should_ match the Rollup ID 🤞🏻 (feel free to make this a method).

Comment (drwpow) [reply packages/astro/src/vite-plugin-build-html/index.ts:145]: To add a little more context, there are 2 types of IDs according to Vite:

1. `url`: `/@fs/…`
2. `id`: `/Users/astro…` (Linux) or `C:/Users/…` (Windows)

The former can be used for the module graph and in the dev server, but not Rollup. The latter can be used for the module graph and Rollup, but not the dev server.

#2052 changed `viteifyURL` to be `url` (“dev“). Should we have some distinction here between “dev” IDs and “build” IDs? Overall this resolution gets a little confusing and would love feedback on how to make it safer / more predictable.

**Comment by natemoo-re** [comment]
Just commenting that this is blocked by an issue in the compiler. I'm working on it and hope to have a new patch out soon!

File: packages/astro/src/vite-plugin-build-html/index.ts
Comment (drwpow) [reply packages/astro/src/vite-plugin-build-html/index.ts:145]: Resolving for simplicity

File: packages/astro/src/vite-plugin-build-html/index.ts
Comment (drwpow) [inline packages/astro/src/vite-plugin-build-html/index.ts:16]: Remove if `assetId` is no longer used below

File: packages/astro/src/vite-plugin-build-html/index.ts
Comment (drwpow) [inline packages/astro/src/vite-plugin-build-html/index.ts:158]:
```suggestion
                  astroAssetMap.set(url, fs.readFile(new URL(resolved)));
```

File: packages/astro/src/vite-plugin-build-html/index.ts
Comment (drwpow) [inline packages/astro/src/vite-plugin-build-html/index.ts:157]:
```suggestion

```

**Review by jonathantneal — APPROVED** [review]
More relative links. Yes!

**Comment by FlipFloop** [comment]
This is still broken for css background image url paths.

**Pull request merged**
